[ENH] Adapter from multiindex to ListDataset

TNTran92 opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look
at our contribution guide: https://github.com/alan-turing-institute/sktime/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
<!--
Example: Fixes #1234. See also #3456.

Please use keywords (e.g., Fixes) to create link to the issues or pull requests
you resolved, so that they will automatically be closed when your pull request
is merged. See https://github.com/blog/1506-closing-issues-via-pull-requests
-->
Partial solution to Issue #2860

#### What does this implement/fix? Explain your changes.
<!--
A clear and concise description of what you have implemented.
-->
Implement an adapter to convert multiindex format to ListDataset in gluon-ts
The method takes as input a pd-multiindex DataFrame, (optional) a list of categorical feature, (optional) the startdate and (optional) frequency of the dataset. 

#### Does your contribution introduce a new dependency? If yes, which one?

<!--
If your contribution does add a new hard dependency, we may suggest to initially develop your contribution in a separate companion package in https://github.com/sktime/ to keep external dependencies of the core sktime package to a minimum.
-->
Dependency: gluon-ts

#### What should a reviewer concentrate their feedback on?

<!-- This section is particularly useful if you have a pull request that is still in development. You can guide the reviews to focus on the parts that are ready for their comments. We suggest using bullets (indicated by * or -) and filled checkboxes [x] here -->

Testing with different dataset in sktime to weed out the bugs. So far, Arrowhead, ItalyPoowerDemand and [StandWalkJump](http://www.timeseriesclassification.com/dataset.php) have been tested.

#### Any other comments?
<!--
Please be aware that we are a loose team of volunteers so patience is necessary; assistance handling other issues is very welcome. We value all user contributions, no matter how minor they are. If we are slow to review, either the pull request needs some benchmarking, tinkering, convincing, etc. or more likely the reviewers are simply busy. In either case, we ask for your understanding during the review process.
-->

#### PR checklist
<!--
Please go through the checklist below. Please feel free to remove points if they are not applicable.
-->

##### For all contributions
- [x] I've added myself to the [list of contributors](https://github.com/alan-turing-institute/sktime/blob/main/.all-contributorsrc).
- [x] Optionally, I've updated sktime's [CODEOWNERS](https://github.com/alan-turing-institute/sktime/blob/main/CODEOWNERS) to receive notifications about future changes to these files.
- [ ] I've added unit tests and made sure they pass locally.
- [x] The PR title starts with either [ENH], [MNT], [DOC], or [BUG] indicating whether the PR topic is related to enhancement, maintenance, documentation, or bug.

##### For new estimators
- [ ] I've added the estimator to the online documentation.
- [x] I've updated the existing example notebooks or provided a new one to showcase how my estimator works.


<!--
Thanks for contributing!
-->

fkiraly reviewed (changes requested):
Thanks!

May I suggest some changes:
* remove dependency on `convert_to` and `check_is`. I'll integrate it into the conversion framework later. Assume the input is `pd-multiindex`.
* do you need to input frequency? Can that not be inferred from the `pandas` input time index? This should be inferred and appropriately translated, instead of being overwritten, in the vanilla call.
* could you move this to `datatypes.adapters`?
* a back-conversion function would be nice, we need that for forecasts, do we not? Maybe a separate PR
* can you add a test please, with example input and outputs?
* can you kindly add

TNTran92 commented:
> * could you move this to `datatypes.adapters`?

I don't see this in the repo. Can you help me locate it?

TNTran92 commented:
> * a back-conversion function would be nice, we need that for forecasts, do we not? Maybe a separate PR

Yes, this will be in a separate PR. That one is in progress.

fkiraly commented:
>I don't see this in the repo. Can you help me locate it?

It does not exist, sorry for being unclear - I suggest creating a new submodule, so we can isolate soft dependencies there.

TNTran92 commented:
> * remove dependency on `convert_to` and `check_is`. I'll integrate it into the conversion framework later. Assume the input is `pd-multiindex`.
> * do you need to input frequency? Can that not be inferred from the `pandas` input time index? This should be inferred and appropriately translated, instead of being overwritten, in the vanilla call.
> * could you move this to `datatypes.adapters`?
> * a back-conversion function would be nice, we need that for forecasts, do we not? Maybe a separate PR
> * can you kindly add tests please, with example input and outputs?

I use `convert_to` because the code works with nested pd.Series. If I remove convert_to, I will have to write my own code somewhere that will do something similar.
Frequency and start_date are now inferred from input data.
the converter has been moved to datatypes.adapters.
Lastly, I am still working on making a pytest...

fkiraly commented:
>Lastly, I am still working on making a pytest...

I can help with the testing, there is a conversion testing framework already in `datatypes`. What it needs are examples that are the same in the format - so you could add examples to the `panel._examples` module?

TNTran92 commented:
> > Lastly, I am still working on making a pytest...
> 
> I can help with the testing, there is a conversion testing framework already in `datatypes`. What it needs are examples that are the same in the format - so you could add examples to the `panel._examples` module?

I added 2 examples: 1 for univariate and 1 for multivariate. Also added listdataset into registry. I'm still trying to figure out pytest

TNTran92 commented:
This is for my information only. After adding datatypes example, encounter
`    def test_check_metadata_inference(scitype, mtype, fixture_index):
        """Tests that check_is_mtype correctly infers metadata of examples.
    
        Parameters
        ----------
        scitype : str - scitype of fixture
        mtype : str - mtype of fixture
        fixture_index : int - index of fixture tuple with that scitype and mtype
    
        Raises
        ------
        RuntimeError if scitype is not defined or has no mtypes or examples
        AssertionError if example metadata is not correctly inferred
        error if check itself raises an error
        """
        # retrieve fixture for checking
>       fixture, _, expected_metadata = get_examples(
            mtype=mtype, as_scitype=scitype, return_metadata=True
        ).get(fixture_index)
E       TypeError: cannot unpack non-iterable NoneType object

fixture_index = 1
mtype      = 'listdataset'
scitype    = 'Panel'

sktime/datatypes/tests/test_check.py:153: TypeError

fkiraly reviewed (changes requested):
@TNTran92, I´ve thought about this.
Given the complexity of some design questions, see #2957, I would suggest:

Let´s separate this PR from the `datatypes` testing framework., i.e., split off the `_examples` and possibly `_registry`.

I would recommend to:
* put the examples and tests in a sub-folder of `_adapters`. Don´t try to integrate with the current test system, just make new tests with expected input/output pairs.
* then, focus next on the back-conversion, while we think about #2957 and integration with the current testing framework (the soft dep question needs to be addressed first).

That would prevent us from getting stuck, and allows @AurumnPegasus to use the conversions already for `gluonts` neural networks.

(this is nothing related to the quality of your work - great work! This is only about removing a conditionality on integration to give @AurumnPegasus access to the functionality, and delaying the integration to a later piece of work)

TNTran92 commented:
> 



> @TNTran92, I´ve thought about this. Given the complexity of some design questions, see #2957, I would suggest:
> 
> Let´s separate this PR from the `datatypes` testing framework., i.e., split off the `_examples` and possibly `_registry`.
> 
> I would recommend to:
> 
> * put the examples and tests in a sub-folder of `_adapters`. Don´t try to integrate with the current test system, just make new tests with expected input/output pairs.
> * then, focus next on the back-conversion, while we think about #2957 and integration with the current testing framework (the soft dep question needs to be addressed first).
> 
> That would prevent us from getting stuck, and allows @AurumnPegasus to use the conversions already for `gluonts` neural networks.
> 
> (this is nothing related to the quality of your work - great work! This is only about removing a conditionality on integration to give @AurumnPegasus access to the functionality, and delaying the integration to a later piece of work)

Understood. I will try to get the unit test and the back-conversion done as soon as possible.

TNTran92 commented:
I have merge conflict that can't be resolved on this branch, so I will be closing this pull request and add the unit test plus everything done so far to the new PR
